### PR TITLE
framework: add initready and init method

### PR DESCRIPTION
**Pull request by MouceL**
use InitReady() to indicate whether the framework is ready

```go
type InitReady interface {
	InitReady() bool
        // InitReady() <-chan struct
}

type ConfigController interface {
	viewstore.ViewerStore
	RegisterEventHandler(kind resource.GroupVersionKind, handler func(resource.Config, resource.Config, Event))
	Run(stop <-chan struct{})
	InitReady
}
```

submodule can determine whether ready as following

```go
func (m *Module) Setup(opts module.ModuleOptions) error {
	cbs := opts.InitCbs
	cbs.AddStartup(func(ctx context.Context) {
		go func() {
                        if env.ConfigController != nil {
                                env.WaitReady(env.ConfigController)
                         }

			// Create the server for the discovery service.
			registryServer, err := server.NewServer(&server.Args{
				SlimeEnv:     opts.Env,
				RegistryArgs: regArgs,
			})
			if err != nil {
				log.Errorf("failed to create discovery service: %v", err)
				return
			}

			// Start the server
			if err := registryServer.Run(ctx.Done()); err != nil {
				log.Errorf("failed to start discovery service: %v", err)
				return
			}
		}()
	})
}	
```

### CODE REVIEW

File: framework/model/module/module.go [hunk 1]
```diff
@@ -83,6 +84,7 @@ type ModuleOptions struct {
 }
 
 type Module interface {
+	Init(opts ModuleOptions) error
```
Comment (YonkaFang) [inline framework/model/module/module.go:87]: why we need `ModuleOptions` for both init and run/setup?

Comment (YonkaFang) [reply framework/model/module/module.go:87]: also we should have comments to explain the difference of these two methods.

File: staging/src/slime.io/slime/modules/lazyload/module/module.go [hunk 2]
```diff
@@ -47,30 +66,17 @@ func (mo *Module) InitScheme(scheme *runtime.Scheme) error {
...
-	}
-	sfReconciler := controllers.NewReconciler(
-		controllers.ReconcilerWithCfg(&m.config),
-		controllers.ReconcilerWithEnv(env),
-		controllers.ReconcilerWithProducerConfig(pc),
-	)
-	sfReconciler.Client = mgr.GetClient()
-	sfReconciler.Scheme = mgr.GetScheme()
-
 	opts.InitCbs.AddStartup(func(ctx context.Context) {
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/lazyload/module/module.go:77]: why we still add callbacks in `setup`/`run`?

File: framework/bootstrap/configcontroller.go [hunk 3]
```diff
@@ -16,145 +21,109 @@ import (
...
-
 	log.Infof("start IstioConfigController successfully")
-
 	return cc, nil
 }
 
-func NewConfigController(config *bootconfig.Config, cfg *rest.Config) (ConfigController, error) {
-	var cc *configController
-	stopCh := make(chan struct{})
+func NewConfigController(configSources []*bootconfig.ConfigSource, stop <-chan struct{}) (*configController, error) {
```
Comment (YonkaFang) [inline framework/bootstrap/configcontroller.go:70]: I can't get why we change this.

File: staging/src/slime.io/slime/modules/lazyload/module/module.go [hunk 4]
```diff
@@ -47,30 +66,17 @@ func (mo *Module) InitScheme(scheme *runtime.Scheme) error {
...
 }
 
-func (mo *Module) Clone() module.Module {
-	ret := *mo
+func (m *Module) Clone() module.Module {
+	ret := *m
 	return &ret
 }
 
 func (m *Module) Setup(opts module.ModuleOptions) error {
```
Comment (YonkaFang) [inline staging/src/slime.io/slime/modules/lazyload/module/module.go:75]: the naming of `InitCbs` seems to be not that accurate.

File: framework/model/module/module.go [hunk 5]
```diff
@@ -424,19 +422,60 @@ func Main(bundle string, modules []Module) {
...
+			InitCbs:           cbs,
+			Manager:           mgr,
+			LeaderElectionCbs: le,
+		}); err != nil {
+			log.Errorf("mod %s init met err %v", mc.env.Config.Name, err)
+			fatal()
+		}
+	}
+
+	// run ConfigController
```
Comment (YonkaFang) [inline framework/model/module/module.go:438]: I'm afraid that the `configController` may emit events when it's not ready and those event-handler modules which depend on `configController` can not handle events correctly then.

File: framework/model/module/module.go (hunk 1 above)
Comment (MouceL) [reply framework/model/module/module.go:87]: only setup is needed

setup: init data struct and register runable functions

File: framework/bootstrap/configcontroller.go (hunk 3 above)
Comment (MouceL) [reply framework/bootstrap/configcontroller.go:70]: split into two parts: NewConfigController and RunController

prepare data struct in NewConfigController

run controller in RunController

File: framework/model/module/module.go (hunk 5 above)
Comment (MouceL) [reply framework/model/module/module.go:438]: These eventhandlers mainly convert the cfg into cache

File: staging/src/slime.io/slime/modules/lazyload/module/module.go (hunk 2 above)
Comment (MouceL) [reply staging/src/slime.io/slime/modules/lazyload/module/module.go:77]: same with above

**Pull request merged**